GTScript-Parallel-model.md

```diff
@@ -134,17 +180,233 @@ The computation domain of every statement is extended to ensure that any require
+    if my_config_var:
+        a = 1
+        b = 2
+    else:
+        a = 2
+        b = 1
+```
+
+translates to:
+
+```python
+for k in range(k, K):
+    parfor ij:
+        if my_config_var:
+            a[i, j, k] = 1
+    parfor ij:
+        if my_config_var:
+            b[i, j, k] = 2
+    parfor ij:
+        if not my_config_var:
+            a[i, j, k] = 2
+    parfor ij:
+        if not my_config_var:
+            b[i, j, k] = 1
+```
+
+### Conditionals on field expressions
+
+- The condition is evaluated for all gridpoints and stored in a mask.
+- Each statement inside the if and else branches is executed according to the same rules as statements outside of branches.
+- Inside the if and else blocks the same field cannot be written to **and** read with an offset in the parallel axises (order does not matter).
+
+**Example for conditionals on field expressions**
+
+```python
+with computation():
+    with interval(...):
+        if field:
+            a = 1
+            b = 2
+        else:
+            a = 2
+            b = 1
+```
+
+translates to:
+
 ```python
 for k in range(start, end):
-    parfor [i_start-2:i_end+1, j_start-1:j_end+2]:
-        u[i, j, k] = 1
-    parfor [i_start:i_end, j_start:j_end]:
-        b[i, j, k] = u[i-2,j,k] + u[i+1,j,k] + u[i,j-1,k] + u[i,j-2,k]
+    parfor ij:
+        mask[i,j] = field[i,j,k]
+    parfor ij:
+        if mask[i,j]:
+            a[i, j, k] = 1
+    parfor ij:
+        if mask[i,j]:
+            b[i, j, k] = 2
+    parfor ij:
+        if not mask[i,j]:
+            a[i, j, k] = 2
+    parfor ij:
+        if not mask[i,j]:
+            b[i, j, k] = 1
+```
+
+The following cases are illegal:
+
+```python
+with computation():
+    with interval(...):
+        if field:
+            b = a[1, 0, 0] # read with offset in 'I' from updated field 'a'
+            a = 1
+            c = a[0, 1, 0] # read with offset in 'J' from updated field 'a'
+
+with computation():
+    with interval(...):
+        if field:
+            a = 1
+        else:
+            b = a[1, 0, 0] # read with offset in 'I' from updated field 'a'
+
+with computation(...):
+    with interval(...):
+        if field:
+            a = a[0, 1, 0] # self assignment with offset (i.e. a read with offset and write)
+```
+
+## Off-center writes in `K`
+
+The vertical direction, `K`, is special in weather and clinate applicaitons. It is used nut just for stencil computations but also for implicit skemes and other algorthms in physical parametrization.
+This is the reason why the vertical look is kept separate from tne horizontal IJ look in the parallel model.
+
+The vertical loop can have _directions_, indicated by the iteraiton policies FORWARD, BACKWARD, or can left unspecified, intending that it could be possibly parallelized. To implement certain skemes it is natural to allow for off-center writes in the vertical direction, since the programmer can know in which order the values are accessed.
```

Comment:
```suggestion
The vertical loop can have _directions_, indicated by the iteraiton policies FORWARD, BACKWARD, or can be left unspecified, intending that it can potentially be parallelized. To implement certain schemes it is natural to allow for off-center writes (e.g. `field[K-1] = ...` instead of `field = ...` in the vertical direction, since the programmer can know in which order the values are accessed.
```